includes/sanitizers/class-amp-img-sanitizer.php

```diff
@@ -17,54 +17,90 @@ class AMP_Img_Sanitizer extends AMP_Base_Sanitizer {
 	private static $script_slug = 'amp-anim';
 	private static $script_src = 'https://cdn.ampproject.org/v0/amp-anim-0.1.js';
 
-	public function sanitize() {
-		$nodes = $this->dom->getElementsByTagName( self::$tag );
-		$num_nodes = $nodes->length;
-		if ( 0 === $num_nodes ) {
-			return;
-		}
-
-		for ( $i = $num_nodes - 1; $i >= 0; $i-- ) {
-			$node = $nodes->item( $i );
-			$old_attributes = AMP_DOM_Utils::get_node_attributes_as_assoc_array( $node );
-
-			if ( ! array_key_exists( 'src', $old_attributes ) ) {
-				$node->parentNode->removeChild( $node );
-				continue;
-			}
-
-			$new_attributes = $this->filter_attributes( $old_attributes );
-
-			// Try to extract dimensions for the image, if not set.
-			if ( ! isset( $new_attributes['width'] ) || ! isset( $new_attributes['height'] ) ) {
-				$dimensions = AMP_Image_Dimension_Extractor::extract( $new_attributes['src'] );
-				if ( is_array( $dimensions ) ) {
-					$new_attributes['width'] = $dimensions[0];
-					$new_attributes['height'] = $dimensions[1];
-				}
-			}
-
-			// Final fallback when we have no dimensions.
-			if ( ! isset( $new_attributes['width'] ) || ! isset( $new_attributes['height'] ) ) {
-				$new_attributes['width'] = isset( $this->args['content_max_width'] ) ? $this->args['content_max_width'] : self::FALLBACK_WIDTH;
-				$new_attributes['height'] = self::FALLBACK_HEIGHT;
-
-				$this->add_or_append_attribute( $new_attributes, 'class', 'amp-wp-unknown-size' );
-			}
-
-			$new_attributes = $this->enforce_sizes_attribute( $new_attributes );
-
-			if ( $this->is_gif_url( $new_attributes['src'] ) ) {
-				$this->did_convert_elements = true;
-				$new_tag = 'amp-anim';
-			} else {
-				$new_tag = 'amp-img';
-			}
-
-			$new_node = AMP_DOM_Utils::create_node( $this->dom, $new_tag, $new_attributes );
-			$node->parentNode->replaceChild( $new_node, $node );
-		}
-	}
+	public function sanitize()
+    {
+        $nodes = $this->dom->getElementsByTagName(self::$tag);
+        $have_dimensions = array();
+        $need_dimensions = array();
+
+        $num_nodes = $nodes->length;
+
+        if (0 === $num_nodes) {
+            return;
+        }
+
+        for ($i = $num_nodes - 1; $i >= 0; $i--) {
+            $node = $nodes->item($i);
+
+            if (!$node->hasAttribute('src')) {
+                $node->parentNode->removeChild($node);
+                continue;
+            }
+
+            // Determine which images need their dimensions determined/extracted
+            if ( !$node->hasAttribute('width') || ! $node->hasAttribute('height')) {
+                $need_dimensions[ $node->getAttribute('src') ][] = $node;
+            } else {
+                $have_dimensions[ $node->getAttribute('src') ][] = $node;
+            }
+        }
+        $this->determine_dimensions($need_dimensions);
+
+        //Calling function twice to save overhead of merging arrays just to send single array to function
```

Comment:
The comment precedes the invocation of the same function twice, each time with a different array as its argument. My first thought was to merge the two arrays and call the function only once to make the code more tidy, but I realized it was not worth the overhead of merging the two arrays. I left the comment in case anyone else had the same idea. I'll remove it, though, since it probably only creates confusion.